doc/reference.md

```diff
@@ -0,0 +1,509 @@
+}
+```
+
+Jsonnet objects can be indexed either using `.` and an identifier or `[]` with an arbitrary expression.
+
+```
+local obj = {
+  "foo": 1,
+  "bar": {
+    "arr": [1, 2, 3],
+    "number": 10 + 7,
+  }
+};
+[
+  obj.foo,
+  obj["foo"],
+  obj["f" + "oo"]
+]
+```
+
+### Inheritance
+
+Jsonnet objects allow inheritance in the OOP sense, even though there are no classes or declarations. The inheritance is realized as an operation `+` which can be applied to any two objects. This might be surprising, because in mainstream languages the inheritance hierarchy is static.
+
+For objects which are simple key-value mappings, inheritance is the same as replacing the respective fields from the first object with fields from the second object.
+
+```
+{
+  a: 1
+  b: 2,
+}
++
+{
+  a: 3
+}
+```
+
+It is possible for one field to refer to other ones using `self` and it is possible to refer to the fields of the base object using `super`.
+
+```
+local obj = {
+  name: "Alice",
+  greeting: "Hello, " + self.name,
+};
+[
+  obj,
+  obj + { name: "Bob" },
+  obj + { greeting: super.greeting + "!"}
+]
+```
+
+In general, it is useful to think about Jsonnet object as a stack of *layers*. A layer consists of fields. An object literal or an object comprehension is a single-layer object. Object inheritance `A` + `B` results in an object with `B` layers put on top of `A` layers. 
+
+Reference to a field through `self` corresponds to looking for the field starting from the top of the stack and going towards the bottom until the field is found. Reference through `super` searches for the field starting from the layer below the current one.
+
+It is easy to emulate functions using objects. It goes without saying that it is not a good style to use in the real code. 
+
+```
+local add = {
+  params: {
+    a: error "please provide argument a",
+    b: error "please provide argument b",
+  },
+  result: self.params.a + self.params.b
+};
+(add + { params: { a: 1, b: 2} }).result
+```
+
+#### Properties
+
+Let `D`, `E`, `F` range over arbitrary expressions evaluating to objects. Let ≡ mean equivalence.
+
+* **Associativity** always holds:
+  ```
+  (D + E) + F   ≡   D + (E + F)
+  ```
+* **Identity** always holds:
+  ```
+  D + { }   ≡   D
+  { } + D   ≡   D
+  ```
+* **Idempotence** holds when `D` does not contain `super`:
+  ```
+  D + D ≡ D
+  ```
+* **Commutativity** holds when `D` and `E` do not contain super and have no common fields:
+  ```
+  D + E   ≡   E + D
+  ```
+
+### Self-Referencing Objects
+
+Objects can be self-referencing even without OOP features, simply because variable definitions can be recursive:
+
+```
+local obj = {
+  name: "Alice",
+  greeting: "Hello, " + obj.name,
+}; obj
+```
+
+Referencing `obj` like that is **different** from using `self`. Here `obj` is a fixed object with a fixed set of fields and their values, and consequently `obj.name` is fixed. On the other hand, `self` is the "current" object, which may be a result of inheritance.
```

Comment:
I think it's worth tying this into your stack analogy.  obj is the current level in the stack, whereas self is the top (bottom?) of the stack, however it is arranged later.

Reply:
You are both kinda right, but not exactly.

> obj is the current level in the stack

Not exactly. It's the top of some fixed stack (fixed object). This is not the same as starting at the current level of the current object.

```
local obj = {
    a: obj.b,
    b: "c" in super,
};
{ c: "whatever" } + obj
```

The result is:
```
{
  a: false,
  b: true,
  c: whatever
}
```

> whereas self is the top (bottom?) of the stack, however it is arranged later.

Yes, top.

> self can reference a field in the current level of the stack or any layer below it

It always starts at the top. This is how overriding works.

> while obj will be looked up as any other variable name (i.e. in the lexical scope).

Yes.

Reply:
I've added a couple paragraphs which should clarify this a little.